Exo 1 solution

GregoireHENRY opened the pull request:
+ setup field from constants within the script
+ plant several trees
+ randomly
+ up to a maximum empiric number
+ handling perspective

list of TODOs in the script

GregoireHENRY committed:
Initial setup for the solution of the 1st exercize

+ setup constants
+ prepare print
+ list of TODOs

GregoireHENRY committed:
Place a random trunc

+ get a single random place for a tree in the field
+ place a trunk at this spot

GregoireHENRY committed:
Place several random trees

+ plant whole trees
+ place several trees

GregoireHENRY committed:
Added perspective

+ changed max number of tree with an empiric formula
+ handling perspective by ordering list of places

pymaldebaran reviewed (changes requested):
Alors juste pour ce commit là il y a plusieurs truc à changer :

1. C'est un script python... donc tu dois avoir un Shebang en première ligne. Tu peux regarder mes commentaires sur les premier commit de Yoshimi à ce sujet.
2. On a ici un problème de documentation : à la ligne 2 on a une docstring du fichier avec juste un titre mais ça dit pas ce que fait ton programme. Et ligne 7 on a une docstring qui sert à rien puisque juste mettre "main" en docstring du main... clairement il vaut mieux rien mettre. Donc à l'un ou l'autre de ces emplacement tu doit mettre un vrai documentation qui dit ce que tu fait dans ce fichier.
3. En python les constantes doivent être en `MAJUSCULE_AVEC_DES_UNDERSCORE`. Ici tu as plein de constante que tu note en minuscule donc on supose que c'est des trucs qui vont être modifiés après... alors que non : `width`, `height`, `tree`, etc
4. Ta façon d'initialiser la constante `field` est très belle... mais absolument illisible donc si il y a un bug c'est là qu'il sera et il a des chances de passer inaperçu. Alors comment changer ça ? L'idée de faire une list comprehension est top par contre ces ce que tu met dedans qui est problématique : 
  - `"#"` devrait être une constante genre `BORDER_CHAR`
  - `" "` devrait être une constante genre `EMPTY_CHAR`
  - `(0, width - 1)` là c'est typique, en fait tu veux dire exprimer "les bornes min et max en x du champ" sauf que à la place on a des truc cryptique avec des "magic numbers" cad des nombre en plein milieu du calcul dont on sait pas ce qu'il représentent :point_right: `X_MIN = 0` et `X_MAX = WIDTH - 1` ou `X_MIN_MAX = (0, WIDTH - 1)` c'est toi qui voitce que tu préfère. idem en y :wink: 
5. Ligne 18 à 23 tu as marqué ta TODO list... je trouve pas que ça soit le meilleur endroit pour ça. L'idéal est de créer un fichier `CHANGELOG.md` ou tu utilises la section _[unreleased]_. Pour les détails je te renvoie à la merveilleuse page qui te dit comment faire un CHANGELOG standard [Keep a changelog](https://keepachangelog.com/en/1.0.0/)
6. Il faut toujours avoir un retour à la ligne à la fin du fichier (en fait c'est pour éviter les problème si on traite le fichier en ligne de commande et qu'ensuite on le pipe vers un autre outil #unixstylebaby en effet certains outils standard ne considère comme une ligne que ce qui se termine par `\n` donc ta dernière ligne de code rique d'être oubliée ! (oui c'est con mais c'est une bonne habitude à avoir)

Une fois que tu as corrigé tout ça... ben déjà ce premier commit sera en état d'être validé !

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -2,12 +2,15 @@
     # initialization
     width = 20
     height = 20
+    borders = [4, 4, 5, 1]  # to ensure tree inside field borders
```
Euh tu ne peux pas utiliser une liste pour stocker des x_min, x_max, etc... c'est juste illisible !

Qu'est ce qu'il y a en 0, en 1, en 2, en 3 ???

Il faut idéalement soit créer 4 constantes correctement nommées soit une dataclass ou un namedtuple si tu veux vraiment tout grouper dans une seule variable (ce qui ici me semble inutile).

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -16,12 +19,22 @@ def main():
     # + do perspective (sort by height?)
 
+    number_trees = 1
+    tree_places = list(
```
Cette liste est mal expliquée... il faudrait au minimum un commentaire pour expliquer sa structure : c'est une liste de coordonnées x, y des arbres.

Mais je ne suis meme pas convaincu que ça ait un intérêt de créer cette liste... il vaudrait mieux avoir une liste des x et une liste des y. Puis au moment de ta boucle plus loin tu fait le zip directement (et là on comprend l'intéret du zip en lisant la boucle). En plus ça sera plus optimal car tu utilisera directement le générateur créé par zip sans le stocker dans une boucle (ce qui est lent et prend de la mémoire).

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -16,12 +19,22 @@ def main():
+
+    # plant trees
+    for place_x, place_y in tree_places:
+        field[place_y][place_x] = tree[-1]
```
En core un magic number !!!! Certe le -1 ici veut dire le dernier élément... mais on ne sait pas pourquoi tu met le dernier élément. Tu as le choix : soit tu explique avec un commentaire soit tu met le -1 dans une constante dont le nom permet de comprendre ce que fait ton code...

pymaldebaran reviewed (changes requested):
Bon bien sûr va falloir encore une fois respecter la casse MAJ_AVEC_UNDERSCORE pour tes constantes mais sinon ça va.

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -19,17 +21,23 @@ def main():
+    # + input fixed size as parameters
 
+    # empiric max number of trees for the sake of visibility
+    max_tree_for_field = math.ceil((width * height) ** (1 / 2) / 4)
```
On a besoin d'un commentaire d'explication sur cette formule cabalistique... d'où tu la sort et tout.

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -19,17 +21,23 @@ def main():
+    max_tree_for_field = math.ceil((width * height) ** (1 / 2) / 4)
     # get random number of trees at random places
-    max_tree_for_field = 3  # need to think on this number
     number_trees = random.randint(1, max_tree_for_field)
```
Oh un magic number : c'est quoi ce `1` sorti de nulle part. Peut-être est-ce le nombre minimum d'arbre... mais alors il aurait fallut le mettre dans une constante genre `MIN_TREE_FOR_FIELD`, non ?

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -19,17 +21,23 @@ def main():
         )
     )
+    # sort places by height for overlapping perspective
+    sorted(tree_places, key=itemgetter(1))
```
Tu te prends la tête pour rien... c'est de nombres aléatoires : inutile de trier les x en même temps que les y ! Trie juste les y !

Ainsi tu pourras continuer à avoir 2 listes séparée pour les x et les y comme demander dans mes com' du commit précédent.

Sinon l'autre solution ça serait d'utiliser des namedtuple plutôt que des tuple dans ta liste genre :
```
Coord = namedtuple('x', 'y')
```
Au moins on comprendrait ce que tu fait quand tu manipules le contenu de ta liste. Mais perso j'ai tendance à préférer bosser avec 2 liste dans ce cas précis. C'est toi qui voit.

GregoireHENRY replied:
Ok pour zip après construction de list X et Y si besoin de clareté. Mais peux-tu m'expliquer pourquoi trier une liste de tuple par ordre croissant des seconds élements pose problème ? Si tu ne tries que ta liste Y tu ne gardes pas en mémoire ces changements et la liste X n'est pas triée. Pour moi c'est simple, efficace et clair. Peux-tu m'expliquer ? Je pense que je rate quelque chose là ..

pymaldebaran replied:
Quand tu tires 2 séries de nombre au hasard de même taille... tu obtiens une série de couple dont les valeurs sont aléatoires (captain obvious). Voyons maintenant les 2 cas de figure :

1. Maintenant si tu tries juste la seconde liste puis que tu fait l'appariement 2 à 2 tu obtiens une liste de couple dont les valeurs sont tirées au hasard mais triée selon y.
2. Et si tu apparie puis fait le trie sur les y tu obtiens un liste de couple dont les valeurs sont tirées au hasard mais triée selon y.

Dans les deux cas tu obtiens la même chose. La seule différence c'est que TOI tu auras vu qu'on a pas déplacé les x en même temps que les y... mais on s'en fout c'est des nombres tiré au hasard ! Que tu les apparie avec les y avant ou après ne change rien à leur hasaritude (qui est la seule chose qui compte ici vu qu'ils n'ont aucune propriété).

Or la méthode n°2 est plus complexe que la n°1 donc on ne doit pas l'utiliser. Il faut toujours chercher à réduire la complexité du code.

Je sais que c'est dur à comprendre car on pense que les x et y sont "ensemble" dès le début... alors que bah en fait non ça change rien. Pense ça comme "_je tire des y au hasard mais croissant et je leur associe des x au hasard_".

GregoireHENRY replied:
Merci beaucoup pour la patience pour l'explication. C'est clair maintenant !

pymaldebaran reviewed (changes requested):
Une fois encore pense à écrire tes constantes correctement.
Et vire cette TODO qui traine au milieu du code :point_right: zou dans le CHANGELOG (voir mes com' précédents)

GregoireHENRY commented:
Je ne vois aucun endroit pour répondre à ton premier commentaire, celui qui commence par :

> Alors juste pour ce commit là il y a plusieurs truc à changer :
> 
> 1. C'est un script python... donc tu dois avoir un Shebang en première ligne. Tu peux regarder mes commentaires sur les premier commit de Yoshimi à ce sujet.

Mes questions suite à tes retours :

+ Pylint refuse les constantes en majuscules définies dans une fonction. Dois-je vivre avec ce warning et devoir forcer mes config `.pre-commit` ou laisser un marquage `disable` ? Dois-je les définir en dehors du main ? Et si je les voulais vraiment dans le main ?
+ Black m'empêche la ligne vide en fin de fichier, et d'autres IDE avec leurs fixers de mémoire aussi. Dois-je vraiment le faire ?
+ Ok, j'améliore la docstring du fichier. Mais comment faire la docstring de la fonction main ? Si je l'enlève (car clairement inutile et redondant de la docstring du fichier dans cet exemple) pylint râle encore.

GregoireHENRY committed:
Added changelog

GregoireHENRY committed:
Shebang and docstring improved

GregoireHENRY committed:
Moved constants outside of function

GregoireHENRY committed:
Removed magic numbers, added constant names

GregoireHENRY committed:
Moved inlined TODO to CHANGELOG

GregoireHENRY committed:
Details on an empiric formula

GregoireHENRY committed:
Fixed perspective issue

Sorted list was not assigned

GregoireHENRY commented:
La seule partie que je n'ai pas corrigée est la création de position X et Y random en attendant tes conseils.

pymaldebaran commented:
> Je ne vois aucun endroit pour répondre à ton premier commentaire, celui qui commence par :
> 
> > Alors juste pour ce commit là il y a plusieurs truc à changer :
> > 
> > 1. C'est un script python... donc tu dois avoir un Shebang en première ligne. Tu peux regarder mes commentaires sur les premier commit de Yoshimi à ce sujet.
> 
> Mes questions suite à tes retours :
> 
>     * Pylint refuse les constantes en majuscules définies dans une fonction. Dois-je vivre avec ce warning et devoir forcer mes config `.pre-commit` ou laisser un marquage `disable` ? Dois-je les définir en dehors du main ? Et si je les voulais vraiment dans le main ?
> 
>     * Black m'empêche la ligne vide en fin de fichier, et d'autres IDE avec leurs fixers de mémoire aussi. Dois-je vraiment le faire ?
> 
>     * Ok, j'améliore la docstring du fichier. Mais comment faire la docstring de la fonction main ? Si je l'enlève (car clairement inutile et redondant de la docstring du fichier dans cet exemple) pylint râle encore.

# Le problème de ne pas pouvoir répondre à mon msg

Oui tu as raison j'ai fait une erreur j'ai fait un commentaire sans l'attacher au code... et ceux là on est obligé d'y répondre temporellement, on peut pas faire de reply. My bad. Pour toutes les prochaines fois je ferai seulement de commentaire de code associé au code.

# Pylint et les CONSTANTES

Alors ton problème de Pylint... bah déjà pylint s'utilise toujours avec une config customisée. Ses valeur "par défaut" ne sont pas génériques mais juste spécifiques à l'équipe de pylint. C'est pas comme black qui a été prévu pour former une façon standard de coder qu'on doit respecter. Donc l'argument du "c'est pylint qui veut pas" n'est pas le bon pour plusieurs raisons :

1. Si on utilise pylint sur un projet... il faut se mettre d''accord avec toute l'équipe sur sa config, cette config doit faire partie du git et on doit signaler dans la doc que le code doit passer à travers pylint pour être accèpté. Là tu n'as pas encore discuté de la config de pylint avec les autres membre du projet. On peut l'intégrer plus tard si tu veux mais ça constituera un PR séparé.
2. Si on regarde ce que dit la règle en question de pylint c'est (en gros) : les variables s'écrive en `snake_case` et les constantes doivent s'écrire en `SCREAMING_SNAKE_CASE`. Ce qui semble être ok pour nous... sauf que comment pylint définit une constante ? C'est une variable dont on ne modifie pas la valeur et **définie au niveau du module**. Mais pourquoi donc seulement au niveau du module ? PEP8 : « Constants **are usually defined on a module level** and written in all capital letters with underscores separating words. Examples include `MAX_OVERFLOW` and `TOTAL`. » :point_right: pylint force une règle qui est marqué "are usually" donc "sont généralement". Donc c'est un choix arbitraire de l'équipe de pylint de dire que seule les constante définie au niveau du module sont des constantes. (en fait cette règle est là pour empêcher les gens d'écrire des variables globales).
3. Énormément de programmeur suive la règle "tout ce qui est constant est en `SCREAMING_SNAKE_CASE` car cela permet de voir rapidement les bugs (on modifie une cosntante --> bug !!!!) et surtout ça permet de transcrire facilement la notion de cosntante/immutabilité de nombreux autres langage (C, C++, Rust, Go, Java...).

Alors qu'est-ce que tu fais ?

Eh bah pour le moment si tu veux continuer à utiliser pylint, tu va devoir changer ta config pour coller aux pratique : tu doit changer la regex des variable pour accepter le tout `SCREAMING_SNAKE_CASE` en plus du `snake_case`. Et comme ça tu pourras proposer ton fichier de config comme référence quand on voudra ajouter pylint au projet.

# Black et le retour à la ligne de fin de fichier

Alors je viens de tester : black n'enlève pas les fin de ligne en fin de fichier... mais apparemment intégré à certains éditeur on peut avoir l'impression que oui. Alors je te l'affirme : il ne le fait pas et ne devrait pas le faire. Si ton éditeur le fait... c'est une erreur. Pour info tu as un truc pour vérifier ça via pre-commit : **end-of-file-fixer** - Ensures that a file is either empty, or ends with one newline. 

# Pylint et la docstring de main

Alors, là aussi c'est à la fois une histoire de config de pylint et une histoire de choix : 

- tu dois configurer pylint pour qu'il fasse ce que toi et ton équipe voulez (tu vois l'énorme différence avec black maintenant ?). En l'occurence tu peux utiliser la config `no-docstring-rgx` de pylint : 

> Regular expression which should only match function or class names that do not require a docstring.
>
> Default: ^_

- Soit tu te force à écrire une doc string utile pour ton main et une docstring utile utile pour ton fichier (et sans redite !!!) soit tu choisit que seul l'un des deux est suffisant (et dans ce cas là tu dois config correctement pylint).

pymaldebaran reviewed (dismissed):
Très bien. Super clair.

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -34,10 +34,10 @@
 def main():
     # initialization
-    # create field with borders
+    # create field with BORDERS
```
Attention, ton search and replace a aussi modifié le commentaire : dans le com borders doit rester en minuscules.

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -34,10 +34,10 @@
     field = [
-        ["#" if (j in (0, width - 1) or i in (0, height - 1)) else " " for j in range(width)]
-        for i in range(height)
+        [BORDER_CHAR if (j in X_LIMITS or i in Y_LIMITS) else EMPTY_CHAR for j in range(WIDTH)]
```
Grâce à ta réécriture on voit bien que la variable `j` en fait c'est des x et que la variable `i` en fait c'est des y.

Donc il faudrait les nommer comme ça x et y ça rendrait l'algo plus lisible. Et oui il faut résister à toujours appeler son premier indice de boucle `i` et le second `j`, ce sont de mauvaises habitude.

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -41,10 +41,8 @@ def main():
-    # get random number of trees at random places
-    number_trees = random.randint(1, max_tree_for_field)
-    tree_places = list(
+    # WIDTH and HEIGHT sizes contribute to the number of trees
```
Je ne comprends toujours pas pourquoi il y a une _puissance 1/2_ et un divisé par 4. Et au passage il faut utiliser math.sqrt pour faire des racine carré en python...

Donc il va me falloir l'origine de cette formule pour que je puisse dire si elle est juste mathématiquement ou pas.

pymaldebaran replied:
Je ne comprends toujours pas pourquoi il y a une puissance 1/2 et un divisé par 4. Et au passage il faut utiliser math.sqrt pour faire des racine carré en python...

Donc il va me falloir l'origine de cette formule pour que je puisse dire si elle est juste mathématiquement ou pas.

GregoireHENRY committed:
Fixed typo in module docstring

GregoireHENRY committed:
Changed list comprehension indexed iterators

GregoireHENRY committed:
Changed method to get tree random positions

+ x and y list separately
+ only y sorted
+ updated for iterator in accordance

GregoireHENRY committed:
Changed random range call (to go with prev commit)

GregoireHENRY committed:
Changed max number of tree formula

+ more simple to go to essential
+ more comments
+ quick fix function comment

GregoireHENRY committed:
Remove unused import

GregoireHENRY committed:
Update CHANGELOG.md to v0.3

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -44,22 +44,15 @@ def main():
-    for position_x, position_y in tree_positions:
+    # get number_trees random positions in the field (excluded the border protection area)
+    # list of y positions are sorted to handle perspective by planting them from top to bottom
+    tree_positions_x = random.sample(range(*PLANTABLE_REGION_LIMITS_HORIZONTAL), number_trees)
```
Voilà ce qui se passe quand on commit ligne à ligne : tu n'as pas commité la définition de `PLANTABLE_REGION_LIMITS_HORIZONTAL` et `PLANTABLE_REGION_LIMTIS_VERTICAL` !

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -26,6 +26,8 @@
 PROTECTION_SPAWN_RIGHT = 4
 PROTECTION_SPAWN_TOP = 1
 PROTECTION_SPAWN_BOTTOM = 5
+PLANTABLE_REGION_LIMITS_HORIZONTAL = (PROTECTION_SPAWN_LEFT, WIDTH - PROTECTION_SPAWN_RIGHT)
```
Ah bah voilà nos fameuse définitions ! Tu ne peux pas mettre ça dans un commit séparé... tu si te rend compte après coup que tu as oublié un truc dans ton commit précédent, au lieu d'ajouter un _nouveau commit_ l'historique, tu va amender le précédent :

```shell
git commit --amend
```

Bon si tu t'en rend compte après plusieurs commit c'est plus compliqué...

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -32,19 +32,23 @@
 EMPTY_CHAR = " "
 TREE_DESIGN = ["^", "^ ^", "(o  )", "(o  o )", "U"]
+# empiric factor to reduce tree spawn occurrence
+TREE_OCCURRENCE_FACTOR = 0.25
```
Techniquement ce `TRE_OCCURENCE_FACTOR` a une unité... là c'est des arbre par mètre (trouvé en analysant ta formule suivante). Bon en vrai tu vas voir avec ma prochaine remarque que ça devrait être une constante sans dimension et avec une vraisignification... et un autre nom.

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -32,19 +32,23 @@
+    # but also to have "enough" probability to see a clean field with few trees.
+    # Thus, both WIDTH and HEIGHT sizes contribute to the maximum number of trees, multiplied by
+    # a coefficient TREE_OCCURRENCE_FACTOR to reduce the spawn occurrence of trees.
+    max_tree_for_field = int((WIDTH + HEIGHT) * TREE_OCCURRENCE_FACTOR)
```
Maintenant qu'elle est bien expliquée... je peux voir que cette formule est fausse :smile: 

Tu veux éviter d'avoir trop d'arbre. Mais tout ce que tu sais c'est que `max_tree_for_field` doit te donner ton nombre max d'arbre. Mais ton analyse est fausse... certe ça augmente avec la taille du champ aussi bien en x qu'en y mais dans ce cas là il faudrait un multiplication des deux (si si je te jure c'est une question d'analyse aux dimensions...). En gros c'est proportionnel à la surface.

Mais surtout ça dépend de la surface de ton arbre ! Plus l'arbre prend de place moins tu peux en mettre. Il nous faut donc les dimensions de l'arbre :

```python
TREE_HEIGHT = len(TREE_DESIGN)
TREE_WIDTH = max([len(line] for line in TREE_DESIGN)
```

Maintenant ta formule devient (en majuscule au passage vu que c'est une constante !!!!) :

```python
MAX_TREE_FOR_FIELD = int((WIDTH + HEIGHT) /  (TREE_WIDTH * TREE_HEIGHT))
```

On voit que si l'arbre et le champ font la même taille on ne peut mettre qu'un seul arbre dans le champ en gros. Et niveau dimension ça colle bien : [m]²/[m]² --> nombre. Nickel.

Maintenant comment ajouter un facteur là-dessus pour qu'on puisse contrôler tout ça ? En fait ce qu'on vient de calculer c'est le nombre d'arbre maximum avec une densité de 100% (toute la surface est occupée par des arbre en moyenne). Il suffit de multiplier ça par un nombre entre 0.0 et 1.0 (càd un pourcentage) qui représente le taux d'occupation du sol par les arbre (`TREE_OCCUPATION_RATE`) qui est sans dimension : 

- 0.5 --> la moitié de la surface est couverte en moyenne
- 0.1 --> 10% de la surface est couverte (c'est une prairie avec qq arbres)
- 0.9 --> 90% de la surface est couverte d'arbre (c'est une forêt assez dense)
- 1.5 --> 150% de la surface est couverte donc il y a pus d'arbre que le max, on a forcément des recouvrement dans 50% des cas (en moyenne) c'est une forêt très dense

(:point_up: voilà le genre de commentaire hyper utile à mettre au dessus d'une constante de ce genre : explication de sa signification + exemples clairs avec interprétation)

Ta formule devrait donc être :

```python
MAX_TREE_FOR_FIELD = floor(TREE_OCCUPATION_RATE * (WIDTH + HEIGHT) /  (TREE_WIDTH * TREE_HEIGHT))
```

Tu notera que j'ai utilisé `floor` du module `math` plutôt que `int` car on calcul un nombre maximum : si on a la place pour 12.7 arbres, en nombre entier on a la place pour 12 arbre. Ça doit être explicite.

pymaldebaran commented on Exercice_1_-_Le_champ_de_Noel/champ.py:
```diff
@@ -13,9 +13,7 @@
 The program handles perspective, showing trees "at the bottom" in front of the other trees.
 """
 
-import math
```
Hu hu hu celui-là va revenir... hu hu hu

GregoireHENRY committed:
Change max tree formula

+ max number of trees now depends on field and tree surfaces and
occupation rate
+ explicit upper case naming for constants
